### PR TITLE
Throw error on recursive schemas

### PR DESCRIPTION
**Pull request by Cypher1**
Adds flag to allow testing of recursive schema support and throws error on detection of recursive schemas

Fixes b/172029133

### CODE REVIEW

**Comment by Cypher1** [comment]
CC' @bgogul FYI this adds enforcement of non-recursive schemas (as well as turns on recursive schema support for some tests that already depended on it). Looks like the next bug to fix for recursive schema support will be toLiteral (and similar methods) likely recursing forever when converting them. I'm filing bugs as I find methods that display this behavior.

File: src/runtime/manifest.ts [hunk 1]
```diff
@@ -715,6 +719,32 @@ ${e.message}
...
+            continue;
+          }
+          const fieldType = entityType.getEntitySchema();
+          referenced.add(fieldType.name);
+          visit(fieldType);
+        }
+      };
+      visit(schema);
+      if (referenced.has(schema.name)) {
+        throw new ManifestError(schema.location, `Unsupported recursive schema ${schema.name}`);
```
Comment (csilvestrini) [inline src/runtime/manifest.ts:743]: Suggestion: you could say something like this instead: `Recursive schemas are unsupported, requires enabling flag recursiveSchemasAllowed: ${schema.name}.`

File: src/runtime/flags.ts [hunk 2]
```diff
@@ -24,6 +24,7 @@ class FlagDefaults {
   static useSlandles = false;
   static defaultToSlandles = false;
   static fieldRefinementsAllowed = false;
+  static recursiveSchemasAllowed = false;
```
Comment (csilvestrini) [inline src/runtime/flags.ts:27]: Add a comment explaining what it's for, and link to the tracking bug as well.

File: src/runtime/tests/field-path-test.ts [hunk 3]
```diff
@@ -356,15 +357,15 @@ describe('field path validation', () => {
       resolveFieldPathType(['inlined', 'name'], type);
     });
 
-    it('rejects missing fields nested inside inline schemas', async () => {
+    it('rejects missing fields nested inside inline schemas', Flags.withFlags({recursiveSchemasAllowed: true}, async () => {
```
Comment (csilvestrini) [inline src/runtime/tests/field-path-test.ts:360]: It'd be good to have a few tests to verify that it throws an exception when the flag is disabled

File: src/utils/internal/misc.ts [hunk 4]
```diff
@@ -109,12 +109,16 @@ export function mapToDictionary<T>(map: Map<string, T>): Dictionary<T> {
 
 /** Recursively delete all fields with the given name. */
 // tslint:disable-next-line: no-any
-export function deleteFieldRecursively(node: any, field: string) {
+export function deleteFieldRecursively(node: any, field: string, options: {replaceWithNulls: boolean} = {replaceWithNulls: false}) {
```
Comment (csilvestrini) [inline src/utils/internal/misc.ts:112]: Nit: it's probably better to make this an optional param with `?` instead:

```
options?: {replaceWithNulls?: boolean}
```

and then in the function definition:

```
options = options || {};
```

This makes it easier to extend in the future. If you add another option to `options` in the future, you don't want to have to require specifying `replaceWithNulls` in the options dict, it should be optional too.

**Review by csilvestrini — APPROVED** [review]

File: src/utils/internal/misc.ts (hunk 4 above)
Comment (Cypher1) [reply src/utils/internal/misc.ts:112]: Thanks

File: src/runtime/tests/field-path-test.ts (hunk 3 above)
Comment (Cypher1) [reply src/runtime/tests/field-path-test.ts:360]: Added a few more tests, I think I've covered most cases, should revisit it when new types are added like unions or nullables.

**Pull request closed without merging**